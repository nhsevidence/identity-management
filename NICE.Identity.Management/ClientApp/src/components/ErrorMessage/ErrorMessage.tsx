import React from "react";

import { Alert } from "@nice-digital/nds-alert";

type ErrorMessageProps = {
	error: Error;
	children?: React.ReactNode;
};

export const ErrorMessage = (props: ErrorMessageProps) => (
	<Alert type="error">
		{props.children ? (
			props.children
		) : (
			<>
				<h2>Whoops</h2>
				<p>Sorry, there's been an error!</p>
			</>
		)}
        <dl style={{visibility:"hidden"}}>
            <dt>Message:</dt>
            <dd>{props.error.message}</dd>
            <dt>Stack:</dt>
            <dd>{props.error.stack}</dd>
        </dl>
	</Alert>
);
