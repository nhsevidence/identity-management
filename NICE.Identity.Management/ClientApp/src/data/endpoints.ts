const API_BASE_URL = process.env.REACT_APP_API_BASE_URL
	? process.env.REACT_APP_API_BASE_URL
	: "#{REACT_APP_API_BASE_URL}";

export const Endpoints = {
	usersList: `${API_BASE_URL}/users`, // get all users
	user: (userId: string | number) => `${API_BASE_URL}/users/${userId}`,

	servicesList: `${API_BASE_URL}/services`,
	service: (serviceId: string | number) =>
		`${API_BASE_URL}/services/${serviceId}`,

	userWithRoles: (userId: string | number) =>
		`${API_BASE_URL}/getuserwithroles/${userId}`,
	userRolesByWebsite: (userId: string | number, websiteId: string | number) =>
		`${API_BASE_URL}/users/${userId}/rolesbywebsite/${websiteId}`,

	verificationEmail: `${API_BASE_URL}/VerificationEmail/VerificationEmail`,
	
};
