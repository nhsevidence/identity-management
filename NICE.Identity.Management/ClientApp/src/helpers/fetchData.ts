export const fetchData = async (url: string, options?: {}) => {
	let response, data;
	try {
		response = await fetch(url, options);
		data = await response.json();
	} catch (err) {
		let error: Error = err;
		return error;
	}

	if (response.status === 200) {
		return data;
	} else {
		let error = new Error(data.message);
		return error;
	}
};
