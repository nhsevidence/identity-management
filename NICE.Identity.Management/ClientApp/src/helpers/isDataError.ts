export const isDataError = (data: {} | Error) => {
	if (data instanceof Error) {
		return true;
	}

	return false;
};
