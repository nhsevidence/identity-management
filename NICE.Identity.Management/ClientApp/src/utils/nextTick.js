/**
 *  Returns a new promise that resolves after a 0ms setTimeout, effectively clearing the call stack
 */
export const nextTick = async () => {
	return new Promise(resolve => {
		setTimeout(resolve, 0);
	});
};