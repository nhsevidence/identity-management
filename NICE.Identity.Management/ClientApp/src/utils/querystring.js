import queryString from "query-string";

export function queryStringToObject(qryString) {
	return queryString.parse(qryString);
}

export const removeQueryParameter = (url, parameter, value = null) => {
	const urlParts = url.split("?");

	if (urlParts.length < 2) return url;

	let urlParams = urlParts[1].split(/[&;]/g);

	for (let i = urlParams.length; i-- > 0; ) {
		const parParts = urlParams[i].split("=");

		if (
			parParts[0].toLowerCase() === parameter.toLowerCase() &&
			(!value ||
				(value && value.toLowerCase() === (parParts[1] || "").toLowerCase()))
		) {
			urlParams.splice(i, 1);
		}
	}

	const queryString = urlParams.length > 0 ? "?" + urlParams.join("&") : "";
	return urlParts[0] + queryString;
};

export const stripMultipleQueries = (path, queries) => {
	let strippedPath = path;

	queries.forEach((query) => {
		strippedPath = removeQueryParameter(strippedPath, query);
	});

	return strippedPath;
};

export const appendQueryParameter = (url, parameter, value) =>
	`${url}${url.indexOf("?") === -1 ? "?" : "&"}${parameter}=${value}`;
