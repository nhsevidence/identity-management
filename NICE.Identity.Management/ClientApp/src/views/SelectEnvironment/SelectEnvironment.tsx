import React, { Component } from "react";
import { RouteComponentProps, Link, Switch, Route } from "react-router-dom";

import { Endpoints } from "../../data/endpoints";
import { fetchData } from "../../helpers/fetchData";
import { isDataError } from "../../helpers/isDataError";
import { UserType, ServiceType, EnvironmentType } from "../../models/types";
import { User } from "../User/User";
import { PageHeader } from "@nice-digital/nds-page-header";
import { Grid, GridItem } from "@nice-digital/nds-grid";
import { StackedNav, StackedNavLink } from "@nice-digital/nds-stacked-nav";
import { ErrorMessage } from "../../components/ErrorMessage/ErrorMessage";
import { Breadcrumbs, Breadcrumb } from "@nice-digital/nds-breadcrumbs";

type TParams = { id: string; serviceId: string };

type SelectEnvironmentProps = {} & RouteComponentProps<TParams>;

type SelectEnvironmentState = {
	user: UserType;
	service: ServiceType;
	environments: Array<EnvironmentType>;
	error?: Error;
	isLoading: boolean;
};

export class SelectEnvironment extends Component<
	SelectEnvironmentProps,
	SelectEnvironmentState
> {
	constructor(props: SelectEnvironmentProps) {
		super(props);
		this.state = {
			user: {} as UserType,
			service: {} as ServiceType,
			environments: [],
			isLoading: true,
		};
	}

	async componentDidMount() {
		let user = await fetchData(Endpoints.user(this.props.match.params.id)),
			service = await fetchData(
				Endpoints.service(this.props.match.params.serviceId),
			);

		if (isDataError(user)) {
			this.setState({ error: user });
		}
		if (isDataError(service)) {
			this.setState({ error: service });
		}

		let environments = service.websites;

		this.setState({ user, service, environments, isLoading: false });
	}

	render() {
		const { user, service, environments, error, isLoading } = this.state;
		const { params, url } = this.props.match;
		const { id } = params;

		let nameBreadcrumb = `${user.firstName} ${user.lastName}`;

		if (isLoading) {
			nameBreadcrumb = "Loading user details";
		}
		if (error) {
			nameBreadcrumb = "Error";
		}

		return (
			<>
				<Breadcrumbs>
					<Breadcrumb to="/users" elementType={Link}>
						Users
					</Breadcrumb>
					<Breadcrumb to={`/users/${id}`} elementType={Link}>
						{nameBreadcrumb}
					</Breadcrumb>
					<Breadcrumb to={`/users/${id}/services`} elementType={Link}>
						Select service
					</Breadcrumb>
					<Breadcrumb>Select environment</Breadcrumb>
				</Breadcrumbs>

				{!error ? (
					<>
						<PageHeader
							preheading="Select environment for"
							heading={service.name}
						/>

						<Grid>
							<GridItem cols={8}>
								<StackedNav>
									{environments.map(environment => {
										return (
											<StackedNavLink
												destination={`${url}/${environment.id}/roles`}
												elementType={Link}
											>
												{environment.environmentName}
											</StackedNavLink>
										);
									})}
								</StackedNav>
							</GridItem>
						</Grid>
					</>
				) : (
					<>
						<PageHeader heading="Error" />
						<ErrorMessage error={error}></ErrorMessage>
					</>
				)}
			</>
		);
	}
}
