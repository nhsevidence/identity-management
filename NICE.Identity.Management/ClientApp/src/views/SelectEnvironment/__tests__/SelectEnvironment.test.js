import React from "react";
import { mount, shallow } from "enzyme";
import { MemoryRouter } from "react-router";
import fetchMock from "fetch-mock";
import toJson from "enzyme-to-json";

import { nextTick } from "../../../utils/nextTick";
import singleUser from "./singleUser.json";
import singleService from "./singleService.json";
import { SelectEnvironment } from "../SelectEnvironment";

import * as fetchData from "../../../helpers/fetchData";
import { Endpoints } from "../../../data/endpoints";
import { ErrorMessage } from "../../../components/ErrorMessage/ErrorMessage";

describe("SelectEnvironment", () => {
	const match = {
		params: { id: 1, serviceId: 1 },
		isExact: true,
		path: "",
		url: "",
	};

	afterEach(fetchMock.reset);

	it("should show loading message before data has been loaded", () => {
		fetchMock.get("*", {});
		const wrapper = shallow(<SelectEnvironment match={match} />);
		expect(wrapper.find("p").text()).toEqual("Loading...");
	});

	it("should call fetchData during componentDidMount", () => {
		fetchMock.get("*", {});
		const wrapper = shallow(<SelectEnvironment match={match} />);
		const instance = wrapper.instance();
		jest.spyOn(fetchData, "fetchData");
		instance.componentDidMount();
		expect(fetchData.fetchData).toHaveBeenCalledTimes(1);
	});

	it("should match the snapshot after data has been loaded", async () => {
		fetchMock.get(Endpoints.user(match.params.id), singleUser);
		fetchMock.get(Endpoints.service(match.params.serviceId), singleService);
		const wrapper = shallow(<SelectEnvironment match={match} />);
		await nextTick();
		wrapper.update();
		expect(toJson(wrapper, { noKey: true, mode: "deep" })).toMatchSnapshot();
	});

	it("should show error message when user fetchData function returns 500 error", async () => {
		fetchMock.get(Endpoints.user(match.params.id), 500);
		fetchMock.get(Endpoints.service(match.params.serviceId), singleService);
		const wrapper = mount(
			<MemoryRouter>
				<SelectEnvironment match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when user fetchData function returns 401 error", async () => {
		fetchMock.get(Endpoints.user(match.params.id), 401);
		fetchMock.get(Endpoints.service(match.params.serviceId), singleService);
		const wrapper = mount(
			<MemoryRouter>
				<SelectEnvironment match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when services fetchData function returns 500 error", async () => {
		fetchMock.get(Endpoints.user(match.params.id), singleUser);
		fetchMock.get(Endpoints.service(match.params.serviceId), 500);
		const wrapper = mount(
			<MemoryRouter>
				<SelectEnvironment match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});

	it("should show error message when services fetchData function returns 401 error", async () => {
		fetchMock.get(Endpoints.user(match.params.id), singleUser);
		fetchMock.get(Endpoints.service(match.params.serviceId), 401);
		const wrapper = mount(
			<MemoryRouter>
				<SelectEnvironment match={match} />
			</MemoryRouter>,
		);
		await nextTick();
		wrapper.update();
		expect(wrapper.find(ErrorMessage).exists()).toBe(true);
	});
});
