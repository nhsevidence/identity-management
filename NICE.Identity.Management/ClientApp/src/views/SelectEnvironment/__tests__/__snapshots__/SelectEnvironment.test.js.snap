// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SelectEnvironment should match the snapshot after data has been loaded 1`] = `
<Fragment>
  <Breadcrumbs>
    <Breadcrumb
      elementType={[Function]}
      to="/users"
    >
      Users
    </Breadcrumb>
    <Breadcrumb
      elementType={[Function]}
      to="/users/1"
    >
      Kristin Patrick
    </Breadcrumb>
    <Breadcrumb
      elementType={[Function]}
      to="/users/1/services"
    >
      Select service
    </Breadcrumb>
    <Breadcrumb
      elementType="a"
    >
      Select environment
    </Breadcrumb>
  </Breadcrumbs>
  <PageHeader
    heading="EPPI Reviewer"
    preheading="Select environment for"
  />
  <Grid
    elementType="div"
    gutter="standard"
    horizontalAlignment="left"
    reverse={false}
    verticalAlignment="top"
  >
    <GridItem
      cols={8}
      elementType="div"
    >
      <StackedNav
        aria-label="Environments"
      >
        <StackedNavLink
          destination="/1/roles"
          elementType={[Function]}
        >
          dev
        </StackedNavLink>
        <StackedNavLink
          destination="/2/roles"
          elementType={[Function]}
        >
          test
        </StackedNavLink>
      </StackedNav>
    </GridItem>
  </Grid>
</Fragment>
`;
